### DISCUSSION TITLE
Exposing KafkaConnect Rest API over a route

### DISCUSSION THREAD

**Discussion by cstmgl** [strimzi/strimzi-kafka-operator#4738] [Q&A] [answered]
Hi All,

We've built a visualization tool of the Kafka Connect REST APIs.
It does not do much but shows the configurations, topics, task status, etc...
I want to use this in a different Kubernetes cluster than my kafka so I'm trying to expose this REST API service over a route.
I'm using strimzi version 0.20.1

I have two setups, one using Openshift 3.11 and another using Openshift 4.6
In the one using Openshift 3.11 this works fine but in the one with Openshift 4.6 it does not work.
I believe this is because the cluster operator is generating network policies that prevents access to the KafkaConnect pod.
Anyway I'm trying to find ways to make this access to the service with the REST APIs working over the route but I can't seem to find a way, any ideas?
At the moment the only way I can make this work is inside the same cluster by providing access to the service directly but I would really like to do this over a Route.

**Comment by scholzj** [comment, marked as answer]
If you enabled the Connector operator, then there might be network policies - could that be the case? But you can easily add your own network policy with your own name to enable the access as you want it - including opening it for everyone. (I assume for routes, you might need to give the access to the HA Proxy / OCP Router pods? But right now I do not have OCP cluster around to give it a try and give you the exact network policy).

**Reply by cstmgl** [reply]
Thanks for the feedback.
I get this network policy created by the operator:
```
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  selfLink: >-
    /apis/networking.k8s.io/v1/namespaces/insight-data-flow-nightly/networkpolicies/data-flow-connector-connect
  resourceVersion: '683774222'
  name: data-flow-connector-connect
  uid: 0eddb88a-abf5-4560-99f4-f6a48b35a8d7
  creationTimestamp: '2021-04-10T02:59:47Z'
  generation: 1
  namespace: insight-data-flow-nightly
  ownerReferences:
    - apiVersion: kafka.strimzi.io/v1beta1
      kind: KafkaConnect
      name: data-flow-connector
      uid: d86af8d7-30b5-4b02-87f2-6ca4c134ee66
      controller: false
      blockOwnerDeletion: false
  labels:
    app: data-flow
    app.kubernetes.io/part-of: strimzi-data-flow-connector
    app.kubernetes.io/instance: data-flow-connector
    zone: core-public
    app: data-flow-connector
    strimzi.io/name: strimzi
    app.kubernetes.io/managed-by: strimzi-cluster-operator
    strimzi.io/kind: KafkaConnect
    app.kubernetes.io/name: kafka-connect
    strimzi.io/cluster: data-flow-connector
spec:
  podSelector:
    matchLabels:
      strimzi.io/cluster: data-flow-connector
      strimzi.io/kind: KafkaConnect
      strimzi.io/name: data-flow-connector-connect
  ingress:
    - ports:
        - protocol: TCP
          port: 8083
      from:
        - podSelector:
            matchLabels:
              strimzi.io/cluster: data-flow-connector
              strimzi.io/kind: KafkaConnect
              strimzi.io/name: data-flow-connector-connect
        - podSelector:
            matchLabels:
              strimzi.io/kind: cluster-operator
          namespaceSelector: {}
    - ports:
        - protocol: TCP
          port: 9404
  policyTypes:
    - Ingress
```

I then created this policy to grant the access but it seems to not have worked.
```
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  selfLink: >-
    /apis/networking.k8s.io/v1/namespaces/insight-data-flow-nightly/networkpolicies/data-flow-connector-connect-rest
  resourceVersion: '688551508'
  name: data-flow-connector-connect-rest
  uid: 5075113a-ec8a-4775-b10a-964b3f3241fb
  creationTimestamp: '2021-04-10T02:59:01Z'
  generation: 1
  namespace: insight-data-flow-nightly
  labels:
    app: data-flow
    app.kubernetes.io/component: messaging
    app.kubernetes.io/instance: data-flow-connector-base
    app.kubernetes.io/name: data-flow-connector-base
    app.kubernetes.io/part-of: data-flow
    app.kubernetes.io/version: 1.0.1-dev.1-20abd63
spec:
  podSelector:
    matchLabels:
      strimzi.io/name: data-flow-connector-connect
  ingress:
    - ports:
        - protocol: TCP
          port: 8083
      from:
        - podSelector:
            matchLabels:
              role: connect-rest-access
  policyTypes:
    - Ingress
```

I created that matching label on the pod over the template in the KafkaConnect but somehow I can't make this work

**Reply by scholzj** [reply]
Well, the question is which part does not work ... this:

```yaml
        - podSelector:
            matchLabels:
              role: connect-rest-access
```

should allow access from pods in the same namespace which have label `role` set to `connect-rest-access`. So if you run a pod in the same namespace with this label it should work. Are you saying that does not work? Or are you saying that this does not help you access it through the route?

**Reply by cstmgl** [reply]
I meant that the access to the pod seems to work (i can call it from the other pod by using the service) what does not seem to work is the actual route, because it just cannot reach the service.
I don't understand why but I believe it's because the network policy prevents any access to the pod

**Reply by scholzj** [reply]
So as I tried to explain ... your policy allows the access from the pods with label `role: connect-rest-access` in the same namespace (this is yours). If you wanna access it through a route, you would probably need to either configure the HA Proxy router in the `from` section or just open it completely by using:

```yaml
  ingress:
    - ports:
        - protocol: TCP
          port: 8083
```

**Reply by cstmgl** [reply]
hum understood, I have to test it but I'm worried with the priority of contradicting policies, the one generated by the operator and mine. 
That's why I was also trying to say there could be an access over the pod of connect itself

**Reply by scholzj** [reply]
Having multiple policies should not be an issue. They way they should be evaluated is that once at least one policies allows something then it should be allowed.

**Reply by cstmgl** [reply]
thank you it worked, I don't know why but I was assuming this network policies would be restrictive, meaning if something was preventing access you could only override it to a specific subset of criteria's not just do it like this.
but indeed it worked

**Reply by scholzj** [reply]
TBH, I do not really know the motivation for it ... but it is what it is I guess 🤷🏻‍♂️. Glad it works for you now.
